apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java

```diff
@@ -0,0 +1,171 @@
+        // add unique id to each record
+        PCollection<String> addIds = records.apply("AddUUID", ParDo.of(new DoFn<String, String>() {
+            @ProcessElement
+            public void processElement(ProcessContext c) {
+                String uniqueID = UUID.randomUUID().toString();
+                c.output(uniqueID + "\t" + c.element());
+            }
+        }));
+
+        // convert categorical features to integer according to feature map used in training
+        PCollection<String> preprocess = addIds.apply("Preprocess", ParDo.of(new FeatureMap()));
+
+        // run inference using Deep Java Library
+        PCollection<String> ctr = preprocess.apply("Inference", ParDo.of(new DoFn<String, String>() {
+            @ProcessElement
+            public void processElement(ProcessContext c) throws TranslateException, MalformedModelException, ModelNotFoundException, IOException {
+                Criteria<String, String> criteria = Criteria.builder()
+                        .setTypes(String.class, String.class)
+                        .optTranslator(new CtrTranslator())
+                        .optEngine("MXNet")
+                        .build();
+                try (ZooModel<String, String> model = ModelZoo.loadModel(criteria);
+                     Predictor<String, String> predictor = model.newPredictor()) {
+                    c.output(c.element().split("\t")[0] + "\t" + predictor.predict(c.element()));
+                }
+            }
+        }));
+
+        ctr.apply(TextIO.write().to(options.getOutput()));
+        p.run().waitUntilFinish();
+    }
+
+    public static Map<String, Map<String, String>> readFeatureMap() {
+        Map<String, Map<String, String>> featureMap = new ConcurrentHashMap<>();
+        try (Reader reader = Files.newBufferedReader(Paths.get(System.getProperty("ai.djl.repository.zoo.location") + "/feature_map.json"))) {
+            Type mapType = new TypeToken<Map<String, Map<String, String>>>() {
+            }.getType();
+            featureMap = JsonUtils.GSON.fromJson(reader, mapType);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return featureMap;
+    }
+
+    static class CtrTranslator implements Translator<String, String> {
+        @Override
+        public NDList processInput(TranslatorContext ctx, String input) {
+            String[] fields = input.trim().split("\t");
+            NDManager manager = ctx.getNDManager();
+            float[] features = new float[fields.length - 1];
+            // first column uuid
+            for (int i = 1; i < fields.length; i++) {
+                features[i - 1] = Float.parseFloat(fields[i]);
+            }
+            return new NDList(manager.create(features));
+        }
+
+        @Override
+        public String processOutput(TranslatorContext ctx, NDList list) {
+            NDArray ctr = list.singletonOrThrow();
+            return String.valueOf(ctr.toFloatArray()[0]);
```

Comment:
If ctr is scalar, you can use
```suggestion
            return String.valueOf(ctr.getFloat());
```